### PR TITLE
Enable typeguard in CI

### PR DESCRIPTION
**Pull request by sklam**
Enable [typeguard](https://github.com/agronholm/typeguard) in CI to perform runtime type checking.

### CODE REVIEW

**Comment by sklam** [comment]
https://github.com/numba/numba/pull/6458/commits/e92fc772b0e17bcbd7a3aba05def107fe9364b0d proves that typeguard CI is working as expected by intentionally failing.
https://github.com/numba/numba/pull/6458/commits/18728a36714463ea6173ae7b774d933ca0747423 fixes that intentional error and CI passes.

**Comment by stuartarchibald** [comment]
Reminder to self to make sure this is added to the optional deps.

**Comment by sklam** [comment]
`typeguard` would a an optional dependency for testing only.

**Comment by stuartarchibald** [comment]
> `typeguard` would a an optional dependency for testing only.

I missed the crucial "in the documentation" part of my reminder sentence :)

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the patch. On inspection it seems like it'll do the right thing. Would be good to document `typeguard`, both as an optional package, an env var, and an entry to the dev docs? Thanks again!

**Comment by EPronovost** [comment]
Is it worth adding a negative test that intentionally calls a function with an invalid argument type under `pytest.raises`?

**Comment by sklam** [comment]
> Is it worth adding a negative test that intentionally calls a function with an invalid argument type under `pytest.raises`?

Good idea. Because we have no way to tell if typeguard is not working currently.

**Comment by stuartarchibald** [comment]
Also, out of interest, can a program "know" if it's running under typeguard? I assume it must be able to some how?

**Comment by sklam** [comment]
> Also, out of interest, can a program "know" if it's running under typeguard? I assume it must be able to some how?

Only by convention from the `runtests.py` using the envvar `NUMBA_USE_TYPEGUARD`: https://github.com/numba/numba/pull/6458/commits/9d6877a25593cc7275d5c53a347f11860036618d#diff-eecfde9e7db9ce81ae6b9487c6c3481e5dc43aa5fc8d17f3ff26ede0c4836338R89

File: docs/source/developer/contributing.rst [hunk 1]
```diff
@@ -207,6 +207,12 @@ in the test runner, there is a ``--log`` flag for convenience::
 
     $ python -m numba.runtests --log
 
+To enable :ref:`runtime type-checking <type_anno_check>`, set environment
```
Comment (stuartarchibald) [inline docs/source/developer/contributing.rst:210]:
```suggestion
To enable :ref:`runtime type-checking <type_anno_check>`, set the environment
```

File: docs/source/developer/contributing.rst [hunk 2]
```diff
@@ -207,6 +207,12 @@ in the test runner, there is a ``--log`` flag for convenience::
...
+variable ``NUMBA_USE_TYPEGUARD=1`` and use the `runtests.py` from the source root
```
Comment (stuartarchibald) [inline docs/source/developer/contributing.rst:211]:
```suggestion
variable ``NUMBA_USE_TYPEGUARD=1`` and use `runtests.py` from the source root
```

File: docs/source/developer/contributing.rst [hunk 3]
```diff
@@ -336,6 +342,23 @@ to make sure. The Numba project's private build and test farm will actually
...
 hardware!
 
+
+.. _type_anno_check:
+
+Type annotation and runtime type checking
+'''''''''''''''''''''''''''''''''''''''''
+
+Numba is slowly gaining type annotations. To facilitate the review of pull
+requests that adds typing annotations incrementally, the test suite uses
```
Comment (stuartarchibald) [inline docs/source/developer/contributing.rst:352]:
```suggestion
requests that are incrementally adding type annotations, the test suite uses
```

File: docs/source/developer/contributing.rst [hunk 4]
```diff
@@ -336,6 +342,23 @@ to make sure. The Numba project's private build and test farm will actually
...
+`typeguard`_ to perform runtime type checking, which helps verify the validity
```
Comment (stuartarchibald) [inline docs/source/developer/contributing.rst:353]:
```suggestion
`typeguard`_ to perform runtime type checking. This helps verify the validity
```

File: docs/source/developer/contributing.rst [hunk 5]
```diff
@@ -336,6 +342,23 @@ to make sure. The Numba project's private build and test farm will actually
...
+of type annotations.
+
+To enable runtime type checking in the test suite, users can use the
```
Comment (stuartarchibald) [inline docs/source/developer/contributing.rst:356]:
```suggestion
To enable runtime type checking in the test suite, users can use
```

File: numba/tests/test_typeguard.py [hunk 6]
```diff
@@ -0,0 +1,31 @@
+"""
+Tests to ensure that typeguard is working as expected.
+Mostly this contains negative test to proof that typeguard can catch errors.
```
Comment (stuartarchibald) [inline numba/tests/test_typeguard.py:3]:
```suggestion
This mostly contains negative tests as proof that typeguard can catch errors.
```

File: numba/tests/test_typeguard.py [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+
+    def test_check_ret(self):
+        with self.assertRaises(TypeError):
+            guard_ret(float(1.2))
+
+    def test_check_does_not_work_with_inner_func(self):
+        def guard(val: int) -> int:
+            return
+
+        guard(float(1.2))
```
Comment (stuartarchibald) [inline numba/tests/test_typeguard.py:31]:
```suggestion
        guard(float(1.2))

if __name__ == '__main__':
    unittest.main()

```

File: numba/tests/test_typeguard.py [hunk 8]
```diff
@@ -0,0 +1,31 @@
+"""
+Tests to ensure that typeguard is working as expected.
+Mostly this contains negative test to proof that typeguard can catch errors.
+"""
+
```
Comment (stuartarchibald) [inline numba/tests/test_typeguard.py:5]:
```suggestion
import unittest
```

File: docs/source/developer/contributing.rst (hunk 1 above)
Comment (stuartarchibald) [inline docs/source/developer/contributing.rst:210]: Should we also tell people how to get/install typeguard and add it to the list of optional packages in the dependency list (this seems like a similar use case to e.g. `graphviz` for CFG debugging)?

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the update, a few minor things to resolve else looks good. I've manually checked the new tests via following the new docs, and can confirm that `typeguard` is picking up that which is tested and the instructions for use valid.

File: docs/source/developer/contributing.rst (hunk 1 above)
Comment (sklam) [reply docs/source/developer/contributing.rst]: yes, i'll add that

File: numba/tests/test_typeguard.py [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+        with self.assertRaises(TypeError):
+            guard_ret(float(1.2))
+
+    def test_check_does_not_work_with_inner_func(self):
+        def guard(val: int) -> int:
+            return
+
+        guard(float(1.2))
+
+if __name__ == '__main__':
```
Comment (stuartarchibald) [inline numba/tests/test_typeguard.py:33]:
```suggestion

if __name__ == '__main__':
```
flake8

Comment (sklam) [reply numba/tests/test_typeguard.py:34]: resolved in https://github.com/numba/numba/pull/6458/commits/0c016ec3b2f03b686277185302ea5b2b1159c967

**Review by stuartarchibald — APPROVED** [review]
Thanks for the fixes. Looks good.

**Pull request merged**